packages/labs/context/src/lib/controllers/context-container.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {ContextCallback} from '../context-request-event.js';
+
+/**
+ * A disposer function
+ */
+type Disposer = () => void;
+
+/**
+ * A tuple of context callback, and its disposer
+ */
+type CallbackRecord<T> = [ContextCallback<T>, Disposer];
+
+/**
+ * A simple class which stores a value, and triggers registered callbacks when the
+ * value is changed via its setter.
+ *
+ * An implementor might use other observable patterns such as MobX or Redux to get
+ * behavior like this. But this is a pretty minimal approach that will likely work
+ * for a number of use cases.
+ */
+export class ContextContainer<T> {
+  private callbacks: Set<CallbackRecord<T>> = new Set();
+
+  private _value!: T;
+  public get value(): T {
+    return this._value;
+  }
+  public set value(v: T) {
+    this.setValue(v);
+  }
+
+  public setValue(v: T, force = false) {
+    let changed = false;
+    if (v !== this._value) {
+      changed = true;
+    }
+    this._value = v;
+    if (changed || force) {
+      this.updateContext();
+    }
+  }
+
+  constructor(defaultValue?: T) {
+    if (defaultValue !== undefined) {
+      this.value = defaultValue;
+    }
+  }
+
+  updateContext = (): void => {
+    this.callbacks.forEach(([callback, disposer]) =>
+      callback(this._value, disposer)
+    );
+  };
+
+  addCallback(callback: ContextCallback<T>, subscribe?: boolean): void {
+    if (subscribe) {
+      const record: CallbackRecord<T> = [
+        callback,
+        () => {
+          this.callbacks.delete(record);
+        },
+      ];
+      this.callbacks.add(record);
```

Comment:
Is there a reason we need to store the unsubscribe function here? The unfortunate result is that despite using a Set here, multiple calls with the same callback are not de-duplicated. This would be the case if the callback itself were stored in the Set.

Reply:
If we don't store the `Disposer` function here, then we'd need to create it in the `updateContext` loop  above to pass out the disposer function to each observer so it can remove itself if the subscription is complete. Unless I'm missing something?

Reply:
We can use a map instead: https://github.com/lit/lit/pull/1955/files#r855421440